### ISSUE TITLE
wraps and patch files should be signed and the public keys shipped with meson

### ISSUE THREAD

**Issue by nirbheek** [mesonbuild/meson#3864]
Meson's wrapdb is currently open to a surface of attack: if the wrapdb server is compromised, the wrap files and the patches can be edited by an attacker and we will never notice.

To mitigate this, we should sign the contents of the wrap file which contains the patch and tarball checksum. Something like this:

```
https://wrapdb.mesonbuild.com/v1/projects/zlib/1.2.11/1/get_wrap
→ zlib.wrap
https://wrapdb.mesonbuild.com/v1/projects/zlib/1.2.11/1/get_wrap_asc
→ zlib.wrap.asc
```

Then, before we use `zlib.wrap` we use the signature in `zlib.wrap.asc` to verify the contents.

**Key management** is going to be interesting, and the best I can think of is:

1. Two master keys: a primary and a secondary public key should be shipped with the Meson sources

2. The secondary key exists in case the primary has to be revoked for some reason

3. A list of public keys that are valid for verifying wraps will be signed with a master key from (1) and stored on wrapdb

4. The list will contain public keys from all wrapdb maintainers (all wrapdb maintainers must have gpg keys)

5. When a wrapdb maintainer makes a new release, it must be accompanied with a signature file for that wrap

6. meson will cache the public key list, and will only update it if a wrap contains a pubkey that is not already in the list

**Comment by sarum9in** [comment]
Just to clarify, the surface of attack only applies to newly downloaded wraps. If user has old wrap they are not affected.
I agree that wraps signature is enough since it transitively includes tarball checksum.

Potential challenges to solve
1. We can't store GPG private keys on wrap server. So the current model of single server that creates wraps is not viable. I will work on tools to sign/upload wraps locally then.
2. I don't think I understand the trust path, could you please explain? Some guess:
    1. I am maintainer of wraps, my key is signed with both `(1)` and `(2)`.
    2. I sign all wraps with my key.
    3. User will download the wrap and will have to build a trust path in GnuPG from `(1)` to my key.
3. I think we need more than 2 root keys to build Web of trust. In ArchLinux as far as I know all public keys are downloaded instead and marked trusted to avoid this problem. See https://www.archlinux.org/master-keys/
4. Just thinking if https://keybase.io might be good idea to use for wrapdb somehow. Not sure how but we might avoid reinventing some wheels. It greatly simplifies key management.

@nirbheek thoughts?

**Comment by nirbheek** [comment]
> We can't store GPG private keys on wrap server. So the current model of single server that creates wraps is not viable. I will work on tools to sign/upload wraps locally then.

Sounds good :+1:

> I don't think I understand the trust path, could you please explain? Some guess:
> 
>     I am maintainer of wraps, my key is signed with both (1) and (2).
>     I sign all wraps with my key.
>     User will download the wrap and will have to build a trust path in GnuPG from (1) to my key.
> 

Yes, exactly. GPG signing of your key with (1) and (2) is not needed, since (1) (or (2) if (1) is revoked) will be used to sign the list of valid public keys, which yields the same trust path as key signing and is more flexible to revocation.

> I think we need more than 2 root keys to build Web of trust. In ArchLinux as far as I know all public keys are downloaded instead and marked trusted to avoid this problem. See https://www.archlinux.org/master-keys/

How does ArchLinux validate the list of trusted public keys? Won't that use a root key of some sort?

> Just thinking if https://keybase.io might be good idea to use for wrapdb somehow. Not sure how but we might avoid reinventing some wheels. It greatly simplifies key management.

I'm not sure I understand how keybase would help here. AFAIK it provides identity management, and in our case we have to handle revocation of trust from individuals, revocation of trust from keys, and revocation of the root key (this last one is handled by key servers).

**Comment by sarum9in** [comment]
> Yes, exactly. GPG signing of your key with (1) and (2) is not needed, since (1) (or (2) if (1) is revoked) will be used to sign the list of valid public keys, which yields the same trust path as key signing and is more flexible to revocation.

*Disclaimer: I am not expert in encryption so I might be wrong about this.*
The problem is that this does not work with the default Web of Trust in GnuPG. By default keys are imported as marginally trusted. You need to have at least 3 marginal keys to sign another key to show it as valid. What ArchLinux does is they maintain their own keyring, independent of the user, it is updated automatically with `archlinux-keyring` package. The package contains the list of master keys which are locally signed on import each time package is installed, and revocation list to revoke old keys also. Developers keys do not have to be signed locally as they are signed with master keys. But there are 5 master keys. So if there is only one signature the individual developer keys will not be considered trusted.

**Comment by nirbheek** [comment]
I don't think the web of trust is a good model to follow here, since trust is handed out by Jussi (who is the central authority), not by distributed consensus. Marginal vs complete trust as a concept is also for degrees of trust of identity, so I don't think that works here either.

**Comment by sarum9in** [comment]
@nirbheek do you have some ideas how you see this working without distributed consensus/WoT model?
Do you think we should implement tooling for maintenance of our own keyring on user machines with non-standard settings of GnuPG?
Do you propose to use some different key distribution/verification mechanism?
Or maybe we should ask users to sign Jussi's key and set it as fully trusted and rely on his signatures for maintainers' keys? This will probably confuse a lot of users not familiar with PGP workflows.

**Comment by nirbheek** [comment]
The mechanism I described does not use the WoT or key signing, and does not require users to do anything different.

**Comment by sarum9in** [comment]
I see, I missed `(3)`, sorry about that.
I think I understand the plan now. Just to clarify we should have a file with public keys distributed, not just their IDs. I will update the bug here on how we can do this with GnuPG.

**Comment by sarum9in** [comment]
FTR dumping how this can be done with GnuPG:
```shell
# Create keyring.gpg and sign it with the master keys.
gpg --export KEY1 >>/full/path/to/keyring.gpg
gpg --export KEY2 >>/full/path/to/keyring.gpg
gpg --detach-sign ...
# Distribute /full/path/to/keyring.gpg somehow via wrapdb.
# Now you can use it to verify signatures without importing anything.
gpg --no-default-keyring --keyring /full/path/to/keyring.gpg --verify foo.wrap.sig
```

**Comment by nirbheek** [comment]
> If we ever revoke a GPG key (by removing it from the list), old signatures will be invalid. Since we verify signatures only on download of the *.wrap this will prevent user verifying the old wraps, unless we re-sign them and update all the signatures.

I guess we can fix this by adding a date field to all wrap files and comparing with the date of revocation.

> If we ever revoke a key, a user will not get an update with if we naively cache the keys. We need to implement robust revocation mechanism, possibly by adding cache expiration time.

Yes, this makes sense.

> Old versions of meson will not be able to interact with new version of wrapdb. I don't think this is addressable in any other way than patching software or having a separate distribution of the root keys, which is the same.

Why is that? I proposed that the wrap signature be in a separate file precisely so that backwards compatibility would continue to work.

> Same problem if the cached version of the key has expired. This we can at least automate with cache refresh if key verification fails.

Yes, that's what I was thinking too.

**Comment by sarum9in** [comment]
> I guess we can fix this by adding a date field to all wrap files and comparing with the date of revocation.

I am afraid the moment we decide to make this anywhere complex we are jumping into the rabbit hole of endless corner cases. In the end we will end up either implementing a full CA or failing, I bet on the latter, as it happens with any home-made security systems.

Since wraps already downloaded should continue to work, it's probably not a big deal. It is also an implicit indicator that the old wrap is no longer supported and user has to review it carefully, so maybe we should just be explicit about this design choice and state that at some point in the future unsupported revisions/versions will no longer pass verification.

> Why is that? I proposed that the wrap signature be in a separate file precisely so that backwards compatibility would continue to work.
I think I wasn't clear here, let me clarify. The case I am talking about is:

You have meson which was released 2 years ago. We changed both root keys since then. You no longer can interact with wrapdb without patching your meson version with up to date root keys.

I believe once wrap is downloaded it's user responsibility to do whatever they want with it and it should not be verified as part of the build process, only on download. So building a software is not a concern, only obtaining new wraps.

**Comment by nirbheek** [comment]
> In the end we will end up either implementing a full CA or failing, I bet on the latter, as it happens with any home-made security systems.

I agree, this is a complex problem and it's probably going to end in maintenance pain unless we keep things simple. Let's set up a workflow where if we ever have to revoke a key, and that key is being used for the newest version of a wrap, we verify the wrap and sign it with a fresh key.

> You have meson which was released 2 years ago. We changed both root keys since then. You no longer can interact with wrapdb without patching your meson version with up to date root keys.

I don't think we have to care too much about this edge-case. Having to revoke/change both the root keys is going to be very unlikely (except w.r.t. expiry, not sure how to handle that yet), and people using 2-year-old Meson probably won't be using wrapdb anyway.

**Comment by jpakkane** [comment]
Some things that popped to mind while reading this.

First of all, whatever scheme we come up with it should work without GPG installed on the user's system. Preferably it should be doable with only Python's standard crypto library. Don't know if that is possible, though.

One possible way of doing this would go something like this (not properly crypto analyzed, could be complete bunk):

- generate an eternal master key that exists only on one machine that is not on the Internet
- generate a signing key that will be used to do the actual signing, generate a signing certificate for it using the master key
 - put the master public key in Meson source code
 - put the public signing key + cert on some public location on wrapdb
 - when a new wrap is generated and added to wrapdb, a secret machine holding the signing key downloads the wrap file, signs it and uploads the signature to wrapdb
 - put a list of revoked certificates in Meson code maybe?
 - if the signing key gets compromised, publish a revocation certificate and generate a new key

I don't have the requisite crypto expertise to reliably design and run this. It would be great if we could get some help from people who do.

**Comment by nirbheek** [comment]
> I don't have the requisite crypto expertise to reliably design and run this. It would be great if we could get some help from people who do.

Yes, as we think more about it, this whole setup is turning out to be more thorny than we thought.

Circling back to the original problem of trusting a single VPS hosting service: once we have a 'dumb' wrapdb without a database, we can setup multiple wrapdb servers and require that Meson always download wraps from two unrelated servers and verify that the contents are the same?

I'm imagining that we get trusted individuals (@ignatenkobrain, @nirbheek, etc) and orgs (GNOME, Freedesktop, Debian, etc) to host 1-2 servers for us, and ensure that we pick two different sources (say, igor's server + one of GNOME's servers) while doing the verification. I own and run two linodes, and I'm happy to host a dumb wrapdb on each. I can talk to freedesktop and GNOME folks about hosting too, if we decide to do this.

**Comment by sarum9in** [comment]
> Circling back to the original problem of trusting a single VPS hosting service: once we have a 'dumb' wrapdb without a database, we can setup multiple wrapdb servers and require that Meson always download wraps from two unrelated servers and verify that the contents are the same?

This doesn't solve the problem of trusting github or whatever source of data these stateless replicas use.

**Comment by sarum9in** [comment]
> First of all, whatever scheme we come up with it should work without GPG installed on the user's system. Preferably it should be doable with only Python's standard crypto library. Don't know if that is possible, though.

@jpakkane I don't think it is if the solution is going to use public key cryptography. Python standard library has hashing, for for any interoperability with GnuPG we need RSA library, which I didn't find in Python crypto.

> when a new wrap is generated and added to wrapdb, a secret machine holding the signing key downloads the wrap file, signs it and uploads the signature to wrapdb

Do we make users simply trust the machine? This creates another single point of attack.

> I don't have the requisite crypto expertise to reliably design and run this. It would be great if we could get some help from people who do.

Unfortunately neither do I. However what we can do is we can start with something relatively simple and based on widely known tools like GnuPG. We should make the verification optional at the start anyway and this might actually attract feedback of people who knows how to do it.

So in my opinion we should try this:
1. Maintainers sign wraps with their own PGP keys.
2. Publish the list of maintainers and their public keys signed by few people with write access to meson.
3. Implement an optional experimental flag in meson to verify wrap on download, which requires `gpg` tool. This will download maintainers public keys and verify wrap against them.
4. Wait and gather feedback. At this stage we might decide to revert the flag or change the behavior. Since we are talking about only interactive mode when user downloads the wrap it should not have any impact on automated meson usage.

@jpakkane @nirbheek what do you think?

**Comment by nirbheek** [comment]
> This doesn't solve the problem of trusting github or whatever source of data these stateless replicas use.

Don't we already sign the tags for wrap releases? The wrapdb servers could verify the tag before sending the wrap to the user. It's ok to require gpg on the wrapdb servers.

**Comment by nirbheek** [comment]
> @jpakkane @nirbheek what do you think?

It seems to me that overall we need to:

1. Make wrapdb stateless (+ cache) so you don't need a database to run it
2. Setup multiple wrapdb instances
3. Setup git tag signature verification

Once this is done, if we then require that `meson wrap` must download from multiple wrapdb instances and verify each one, we have a much better chain of trust? Then we can talk about how to implement end-to-end gpg signing at a later stage.

**Comment by sarum9in** [comment]
> Don't we already sign the tags for wrap releases? The wrapdb servers could verify the tag before sending the wrap to the user. It's ok to require gpg on the wrapdb servers.

We don't. My script just creates releases without binding them to particular commits. Also the problem is not the git source code but zip, which is generated somewhere else and `upstream.wrap` in git repository does not contain checksums for it. Both `upstream.wrap` and `patch.zip` are generated on demand and are not reproducible currently.

> must download from multiple wrapdb instances

I don't see it as a good defense mechanism. I can easily imagine rogue proxy changing all of them and breaking this defense. Again, depends on what thread we defend against. This can be partially mitigated by TLS though.

**Comment by nirbheek** [comment]
> I don't see it as a good defense mechanism. I can easily imagine rogue proxy changing all of them and breaking this defense. Again, depends on what thread we defend against. This can be partially mitigated by TLS though.

As you said, a rogue proxy can't break TLS, and I thought we already required TLS for wrapdb downloads? If not, that's really important and is the first step we should take.
